3rdParty/fuerte/src/H1Connection.cpp

```diff
@@ -619,15 +648,24 @@ void H1Connection<ST>::setTimeout(std::chrono::milliseconds millis, TimeoutType
         auto* me = static_cast<H1Connection<ST>*>(s.get());
         if ((type == TimeoutType::WRITE && me->_writing) ||
             (type == TimeoutType::READ && me->_reading)) {
-          FUERTE_LOG_DEBUG << "HTTP-Request timeout\n";
+          FUERTE_LOG_DEBUG << "HTTP-Request timeout" << " this=" << me << "\n";
           me->_proto.cancel();
           me->_timeoutOnReadWrite = true;
           // We simply cancel all ongoing asynchronous operations, the completion
           // handlers will do the rest.
           return;
-        } else if (type == TimeoutType::IDLE) {
+        } else if (type == TimeoutType::IDLE && !me->_writing & !me->_reading && me->_leased == 0) {
           if (!me->_active && me->_state == Connection::State::Connected) {
-            me->shutdownConnection(Error::CloseRequested);
+            int exp = 0;
+            if (me->_leased.compare_exchange_strong(exp, -1)) {
+              // Note that we check _leased here with std::memory_order_seq_cst
+              // **before** we call `shutdownConnection` and actually kill the
+              // connection. This is important, see below!
+              FUERTE_LOG_DEBUG << "HTTP-Request idle timeout"
+                               << " this=" << me << "\n";
+              me->shutdownConnection(Error::CloseRequested);
+              me->_leased = 0;
```

Comment:
> Note that we check _leased here with std::memory_order_seq_cst **before** we call `shutdownConnection` [...]

Actually, it is more important that this assignment to `_leased` uses at least `std::memory_order_release`. This is necessary to ensure that the compare_exchange in `lease` synchronizes-with this store, thereby ensuring that the subsequent check of `_state` in `lease` sees the value written in `shutdownConnection` (regardless which memory ordering was used there).

That's why I usually prefer store/load calls with explicit memory ordering instead of the (more convenient) operator overloads.
```suggestion
              me->_leased.store(0, std::memory_order_seq_cst);
```

Reply:
I am happy with this change.